File: day23/solution.hs
{-# LANGUAGE NamedFieldPuns, ScopedTypeVariables #-}

import Data.Char (isDigit, digitToInt)
import Data.List (tails, foldl', break, sort, isPrefixOf, cycle, unfoldr)
import Data.List.Split (splitOn)
import Control.Monad (guard)
import Debug.Trace (traceShowId, traceShow)

import Data.Map.Strict (Map)
import qualified Data.Map.Strict as Map

data CupGame = CupGame
  { cupOrder :: Map Int Int -- Vector would probably be much faster!
  , currentCup :: Int
  , maxCup :: Int
  } deriving (Show)

main :: IO ()
main = do
  let input = digitToInt <$> "167248359"
  putStrLn "Part 1:"
  let result = solve input 100
  let (pre, post) = break (== 1) $ showCups result
  putStrLn $ concat $ show <$> (drop 1 $ post ++ pre)

  putStrLn "Part 2:"
  let maxInput = maximum input
  let input2 = input ++ [maxInput + 1 .. 1000000]
  let result = solve input2 10000000
  print $ product $ take 2 $ cupsInOrder $ result {currentCup = 1}

solve :: [Int] -> Int -> CupGame
solve input iterations =
  let
    cups = Map.fromList $ zip input (drop 1 $ cycle input)
    game = CupGame cups 1 (maximum input)
  in
    iterate move game !! iterations


cupsInOrder :: CupGame -> [Int]
cupsInOrder CupGame {cupOrder, currentCup, maxCup} = unfoldr (\c -> (\n -> (n,n)) <$> Map.lookup c cupOrder) currentCup

showCups :: CupGame -> [Int]
showCups game@CupGame {currentCup, maxCup} = take maxCup $ cupsInOrder game

countDown :: CupGame -> [Int]
countDown CupGame {currentCup, maxCup} = take maxCup $ (\i -> 1 + i `mod` maxCup) <$> [currentCup - 2, currentCup - 3 ..]

move :: CupGame -> CupGame
move game@CupGame {cupOrder, currentCup, maxCup} =
  let
    removed = take 3 $ cupsInOrder game
    Just removedNext = Map.lookup (last removed) cupOrder
    dest = head $ dropWhile (`elem` removed) $ countDown game
    Just destNext = Map.lookup dest cupOrder
    newOrder = Map.insert dest (head removed) $ Map.insert (last removed) destNext $ Map.insert currentCup removedNext cupOrder
    Just nextCup = Map.lookup currentCup newOrder
  in
    game {cupOrder = newOrder, currentCup = nextCup}

File: day23/solution.py


input = [int(x) for x in '167248359']

def show(next_cup):
    i = 1
    out = []
    for _ in range(len(next_cup)-2):
        out.append(str(next_cup[i]))
        i = next_cup[i]
    print("".join(out))

def solve(input, iterations):
    # create the next_cup array
    next_cup = [0]*(len(input) + 1)
    for i,x in enumerate(input):
        next_cup[x] = input[(i+1)%len(input)]

    maxCup = max(input)
    current = 1
    for _ in range(iterations):
        current = move(maxCup, next_cup, current)
    return next_cup

def move(maxCup, next_cup, current):
    # removed cups:
    r1 = next_cup[current]
    r2 = next_cup[r1]
    r3 = next_cup[r2]

    # clockwise after removed
    r4 = next_cup[r3]

    # find destination cup
    dest = current - 1
    if dest <= 0:
        dest = maxCup
    while dest in [r1, r2, r3]:
        dest -= 1
        if dest <= 0:
            dest = maxCup

    # clockwise after dest
    dest1 = next_cup[dest]

    # update pointers
    next_cup[current] = r4 # cup after removed cups now follows current
    next_cup[dest] = r1 # first removed cup now follows dest
    next_cup[r3] = dest1 # last removed cup now followed by cup that followed dest

    # the current cup moves one cup clockwise
    return next_cup[current]

# Part 1
next_cup = solve(input, 100)
show(next_cup)

# Part 2
input2 = input
current = max(input) + 1
while current <= 1e6:
    input2.append(current)
    current += 1

next_cup = solve(input2, 10e6)
print(next_cup[1]*next_cup[next_cup[1]])
